### ISSUE TITLE
Create new fine-tuning datasets that use fixed vocab. 

### ISSUE THREAD

**Issue by pdakwale** [OpenNMT/OpenNMT-py#355]
Hi,
I am currently trying to do some fine-tuning experiments for domain adaptation. I am trying to use --train_from option where I initialize with a pre-trained model.
My requirement is to train a vocabulary for a large bitext and use it to process and train over subsets of the large set separately. So I am trying to first use preprocess.py to train a vocabulary and then use preprocess.py again for subset bitexts by providing pre-trained vocab through -src_vocab and -tgt_vocab. Current version for preprocess.py in pytorch version builds a combined vocabulary file (combined.data.vocab.pt) for source and target side. However, the input to preprocess.py requires two separate vocab files for source and target. If the output of a previous vocab training was a single file, i can not provide two separate files.
Please clarify if there is a flag to create separate source and target vocab files that I am missing in pytorch version ? 

Also can anyone clarify if there is a better way to do fine-tuning for a model on a different data set
Thanks
Praveen

**Comment by moonlightlane** [comment]
I think there is an option `share_vocab` in `opt.py` that you could set to `False` to meet your need?

**Comment by pdakwale** [comment]
@moonlightlane Thanks for explaining the option. 
Just to clarify : The default setting is share_vocab = True ? 
And so does the output vocabularies are merged by default ?

**Comment by moonlightlane** [comment]
sorry, my bad; I think `share_vocab` is a flag you set when you want to share the vocab among src and tgt. I think the default is NOT share. It is in line 129-130 in `opts.py`

**Comment by pdakwale** [comment]
Got it. However, my problem remains same. By default, I get a single vocab file from pre-processor.
And the input options for using pre-trained vocabulary while pre-processing are separate for source (-src_vocab) and target (-tgt_vocab). How can I use a pre-trained vocabulary for fine-tuning on new data if I have a single file for pre-trained vocab ?

**Comment by PersianNLPer** [comment]
Hi,
I am going through a similar issue:
Assume that I have built the vocabularies (either one for both source and target, or two different vocabs each for one side).
Now, I want to preprocess another parallel corpus (let's say coming from a new domain) to be used for fine-tuning the model.
As far as I understand, in this case the preprocesspy should load the vocabularies, read the new data and just return the indexed version of the data, without changing the vocabularies.
However, as far as I understand from the code, even though there are options for loading the existing vocabularies (-src_vocab and -tgt_vocab), the code doesn't use them at all, and every time produces a vocabulary specific to the given data.
I think this can be easily done by having a condition which checks if -src_vocab and -tgt_vocab are None or not.
Being None, means that we need to create the vocabulary for that side, otherwise we should bypass the vocabulary creation and directly go to the indexing with the existing vocabulary (which should be loaded from the disk).
Any idea?

**Comment by pdakwale** [comment]
@PersianNLPer Good to know that I am not alone !
I tried to modify the code to bypass vocab creation. However, It seems that the preprocess.py doesn't index the bitext based on the vocabulary created (though I am not sure). It first creates the indexed data set and then the vocabs(Lines 59 to 62). So not sure if the indexing is done on the same vocabs.

**Comment by PersianNLPer** [comment]
I am very new to the OpenNMT-Py, so I am not so sure about the details.
But as far as I understand, the indexing has to be done on the created vocabulary. Otherwise, where the indexes come from?
In the next few days I am going to go through the code and try to understand the details.

**Comment by moonlightlane** [comment]
hmm, something like this is also what I would like to do. No idea yet, sorry! I'll also take a deeper look at the code in the next few days.

I think the majority, if not all, of the preprocessing happens in `IO.py`, and seems to be built on top of`torchtext`.

**Comment by srush** [comment]
Okay, let me try to help. I like the ideas here. 

- We do create a vocab in preprocess. By default we create one for src and one for target. You can pass share_vocab and create a single one for both. This vocab is created only from the training data, but used for validation.

- The vocab is stored separately from the data. You could therefore easily use it again by loading it and passing in new data. 

- We currently don't have an option for doing this. However you are right, that it should be as easy as passing the vocab file to preprocess and telling it not to create a new one. 

- Please try this out and send us a PR! 

it should be something like modifying these lines 

```python
    fields = onmt.IO.ONMTDataset.get_fields(nFeatures)
    print("Building Training...")
    train = onmt.IO.ONMTDataset(opt.train_src, opt.train_tgt, fields, opt)
    print("Building Vocab...")
    onmt.IO.ONMTDataset.build_vocab(train, opt)
```

In your case you want to load fields/vocab from disk first and not call build_vocab afterwards. 

```
 fields = onmt.IO.ONMTDataset.load_fields(
                torch.load(opt.dict))
```

**Comment by pdakwale** [comment]
@srush Thanks for the reply. I have already tried changing the code by loading fields for a pre-trained vocab. However, by doing this for pre-processing, my training becomes unstable with increasing losses. To me it seems there is something inconsistent between the indexing of the new data and that in the (pre-trained) dictionary. Do you know any details about it ?

@all does anyone else had a consistent training output with this change ?

**Comment by moonlightlane** [comment]
@pdakwale are you using pre-trained word embedding? If not, then you might want to check in the `Trainer.py` code, in the first for loop in the `train` function, whether the indices in `tgt` and `src` created in the `batch` object correspond to the actual word tokens in your source and target sequences.

i.e. to get the index matrix, you can do `batch.tgt` or `batch.src`. To convert one index back to word tokens, for the source you can do `batch.dataset.fields['src'].vocab.itos[idx]`, where `idx` is an index (entry) in the `batch.src` matrix. For the target, replace `src` with `tgt` in the above code.

To access the original source or target sequence, you can do `batch.dataset.examples[m].tgt` for target, or replace the last `tgt` with `src` for source. Here `m` is the m-th example, and you can get it from the batch by `batch.indices[idx].cpu().data`, where `idx` is the i-th example in the batch.

If you are using (fixed) pre-trained word embedding, I think it is a bit trickier, since from what I understand, word embeddings are part of the model parameters by onmt design (even if you fix those embeddings, they are still part of the model params). Changing the number of word tokens in the source and target means changing the model after loading the pre-trained model. I am not yet sure how to do this; maybe after loading a pre-trained model, I can create new `Embeddings` object using the new training corpus, replace the `self.embeddings` in the encoder and decoder, and then load the word embedding vectors for the new training corpus? I might investigate into this some time next week.

Also, it seems ONMT lua version now supports dynamic dataset as of today: http://forum.opennmt.net/t/opennmt-v0-9-release/942. It might be what you are looking for? It is implemented in lua not in python and I did not yet have a chance to take a look at it, but you might want to take a look at the code and see if similar implementation could be done in py version of onmt without too much labor.

**Comment by surawinfo** [comment]
I am using the changes below, reloading a model and continuing training looks good. I am currently checking if a training stability issue arises with additional experiments. Will verify in detail tomorrow. The main change goes here `../onmt/IO.py` inside `build_vocab` . In this way i think we can keep the rest of the code intact and the reloading or building new vocabulary will depend only on preprocess `-src_vocab` and `-tgt_vocab`. 

`
    @staticmethod
    def build_vocab(train, opt):

        fields = train.fields

        # load src_vocab if path provided
        if opt.src_vocab is not None:
            print('Loading source vocab. from: ' + opt.src_vocab + '...')
            src = ONMTDataset.load_fields(torch.load(opt.src_vocab))
            fields["src"].vocab = src["src"].vocab
        else:
            fields["src"].build_vocab(train, max_size=opt.src_vocab_size,
                                      min_freq=opt.src_words_min_frequency)
            for j in range(train.nfeatures):
                fields["src_feat_" + str(j)].build_vocab(train)

        # load tgt_vocab if path provided
        if opt.tgt_vocab is not None:
            print('Loading target vocab. from: ' + opt.tgt_vocab + '...')
            tgt = ONMTDataset.load_fields(torch.load(opt.tgt_vocab))
            fields["tgt"].vocab = tgt["tgt"].vocab
        else:
            fields["tgt"].build_vocab(train, max_size=opt.tgt_vocab_size,
                                      min_freq=opt.tgt_words_min_frequency)
`

I more thing, while checking the index and the corresponding words between the reloaded and the initial vocabulary, everything is in order. Please, let me know if I missed anything and it can go to PR. @pdakwale I really liked the DYANMIC dataset feature supported by the lua version, will check it in the coming days.

**Comment by pdakwale** [comment]
For long, I was using an older open-nmt version (downloaded in october) and implemented the DYNAMIC dataset features for my requirement. But I pulled new on Feb 13th and I see that there is still issue with this feature in the latest version. To me it seems that the options 'src_vocab' and 'tgt_vocab' are still not being used in preprocess.py which should provide path for pre-trained  source and target vocabularies. Also, it seems that first the training data is indexed and then the vocabulary is created. Does this mean that a new vocabulary will be created for each new data set ? How can I force the vocabulary to be fixed in the latest version ?

**Comment by vince62s** [comment]
maybe this is it: https://github.com/OpenNMT/OpenNMT-py/pull/577/files

**Comment by pdakwale** [comment]
@vince62s Thanks, but this requires vocabulary in a different text file (one word per line format). It doesn't use the vocab.pt file built by preprocess.py. I want to use the vocab.pt file built on a different data set to preprocess a new training data.

**Comment by Henry-E** [comment]
This can't be too hard to integrate. All you need is to access the torchtext vocab object. 

This thread also doesn't make it apparent that the `-src_vocab` and `-tgt_vocab` options are deprecated and not currently working. 

```python
import argparse
import os
import torch

def main():
    parser = argparse.ArgumentParser(description='tag, filter and create '
                                     'source')
    parser.add_argument('-o', '--output_dir_name', help='output directory')
    parser.add_argument('-i', '--input_file_names', nargs='*', help='wikitext'
                                                            ' sentence files')
    parser.add_argument('-t', '--test', help='test')
    args = parser.parse_args()

    for input_file_name in args.input_file_names:
        vocab_objects = torch.load(input_file_name)
        for vocab_object in vocab_objects:
            if vocab_object[0] == 'src':
                output_file_ext = '.src'
            elif vocab_object[0] == 'tgt':
                output_file_ext = '.tgt'
            else:
                continue
            vocab = list(vocab_object[1].freqs)
            input_file_root = \
                os.path.basename(os.path.splitext(input_file_name)[0])
            output_file_name = os.path.join(args.output_dir_name,
                                            input_file_root + output_file_ext)
            with open(output_file_name, 'w') as out_file:
                out_file.write('\n'.join(vocab))

if __name__ == '__main__':
    main()
```

**Comment by vince62s** [comment]
@srush 
Before making a PR I think we need to make choices here.

do we want onmt-py to use "one word per line" vocab files or just vocab.pt files ?
if the latter we need to revisit the src_vocab and tgt_vocab options to load existing .pt files, right ?
(taking into account the share_vocab when necessary)

IMO, the initial need is a real world one, so we need to fullfill this without hacks.

**Comment by Henry-E** [comment]
There's no need to provide a vocab file to the translate function, so the model file must already contain a source and target vocab list of some sort. Otherwise how would it be able to do the mapping?

Currently going through the translate functions so I might end up answering this question myself

**Comment by jsenellart** [comment]
adding my 2 cents on this thread - following up discussion we had in Paris on this specificity of OpenNMT-py to just create on the fly vocab. I am not sure to understand all of the flow and implementation suggestions, but from my perspective, we do need (exactly like in the other frameworks) the ability to pass text precomputed vocabs - since this is corresponding to real-life workflows. 
What is wrong with the `-src_vocab` and `-tgt_vocab`? and why do we need a different format (even if .pt format is the serialized version, being able to use simple text one will enable any type of external vocab computing)?
@surawinfo's patch looks good and if we want to get a bit further than build or load, we can add what we did in lua - if load vocab is different than model vocab (when retraining), we dynamically merge and adapt lookup tables.

**Comment by Henry-E** [comment]
Perhaps a simple solution would be for `preprocess.py` output source, target and feature vocab files instead of a `vocab.pt` file

**Comment by pdakwale** [comment]
Has this feature been added to the current version ? I have been using the src_vocab and tgt_vocab options to load vocabs from text files. But it seems to be that the indexing of data is not based on these pre-trained vocabs. All I want to do is fine-tune a pre-trained model on a additional data set while keeping the vocab and indexing fixed but i get lower BLEU while doing the same with lua version shows expected improvements

**Comment by Henry-E** [comment]
Yep these option are in the code now. Someone can correct me if I'm wrong but the checkpoint files come with their own vocab objects and the preprocessing step with the vocab text file is to make to make sure anything outside that vocab is converted to unk. One thing to check might be that the `stoi` (string to integer) part of the vocab object are the same in the vocab.pt file as the checkpoint's vocab object. If that makes sense?

**Comment by pdakwale** [comment]
hi @Henry-E 
So I am again having problem with latest pull with the above solution. I created a vocab.pt with preprocess.py with a 800 Arabic-English line pairs dataset. 

**preprocess.py -train_src train.arabic -train_tgt $DATA/train.english -valid_src valid.arabic -valid_tgt valid.english -save_data arabic-english-data**

And I extracted files vocab.src and vocab.tgt using your code above. The source and target files vocab size I get are **23494** and **22837** respectively. When I preprocess again on this data while fixing the vocabs extracted above as below:

**preprocess.py -train_src train.arabic -train_tgt $DATA/train.english -valid_src valid.arabic -valid_tgt valid.english -save_data re-arabic-english-data -src_vocab vocab.src -tgt_vocab vocab.tgt**

It reports following vocab sizes in the log after loading the vocabs:

**tgt vocab size: 22776
src vocab size: 827**

Why the source vocabulary size is reduced so much ? Also training on this preprocessed data gives extremely low accuracy and BLEU, while training without fixed vocabularies gives reasonable score 
Is there a bug in the latest version or am I doing something wrong ?
I simply want to fix the vocabulary while training across different datasets. Is there an easy way to do  this ?
src vocab size: 827

**Comment by Henry-E** [comment]
All I can do is suggest manually inspecting and comparing the contents of the different vocab.pt files. The code I listed above should give some clues about how to load the data in the command line. And then which objects to inspect to see the vocab contents. Please let me know how you get on with this. I haven't tried pretraining / fine tuning in a while but I may return to this soon enough.

**Comment by pdakwale** [comment]
Kind of solved this issue. I think there is possibly a bug in **inputters.py  line 373** when loading the vocabulary, it should use UTF-8 encoding to open the vocab files. Without using UTF-8 most of the words in model vocabulary and pre-trained vocabulary do not match at **line 310 or 312**. Changing to UTF-8 reports vocabulary sizes **22776 and 19855** for target and source respectively which is still not same as size of pre-trained vocabulary but much better than 827. Any idea of what else might be prohibiting exact match in the two vocabs ?

Also, to fine-tune, when reading a pre-trained model using -train_from loads optim parameters from the saved model instead of re-initialization which does not give expected performance for fine-tuning, this issue discussed here 
https://github.com/OpenNMT/OpenNMT-py/issues/768
Following changes suggested by @jbecke gives almost similar performance as lua version but still not exactly same

**Comment by Henry-E** [comment]
Have you tried printing the two vocabs to a txt file and checking the difference using https://www.diffchecker.com/diff

You mentioned you're training an Arabic model. It's there's some issue with UTF-8 like you said. For English python 3 sorted out all of those problems I had with strings but maybe there's still some nuance for Arabic alphabet

**Comment by vince62s** [comment]
#990
